parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java

```diff
@@ -878,11 +880,92 @@ public String getFile() {
     return blocks;
   }
 
-  public void setRequestedSchema(MessageType projection) {
+  private boolean uniqueId(GroupType schema, HashSet<Type.ID> ids) {
+    boolean unique = true;
+    List<Type> fields = schema.getFields();
+    for (Type field : fields) {
+      if (field instanceof PrimitiveType) {
+        Type.ID id = field.getId();
+        if (id != null) {
+          if (ids.contains(id)) {
+            return false;
+          }
+          ids.add(id);
+        }
+      }
+
+      if (field instanceof GroupType) {
+        Type.ID id = field.getId();
+        if (id != null) {
+          if (ids.contains(id)) {
+            return false;
+          }
+          ids.add(id);
+        }
+        if (unique) unique = uniqueId(field.asGroupType(), ids);
+      }
+    }
+    return unique;
+  }
+
+  public MessageType setRequestedSchema(MessageType projection) {
     paths.clear();
-    for (ColumnDescriptor col : projection.getColumns()) {
+    HashSet<Type.ID> ids = new HashSet<>();
+    boolean fileSchemaIdUnique = uniqueId(fileMetaData.getSchema(), ids);
+    ids = new HashSet<>();
+    boolean projectionSchemaIdUnique = uniqueId(projection, ids);
+    MessageType schema = null;
+    // if ids are unique, use id resolution
```

Comment:
Instead of figuring out by parquet-mr internals, did you consider letting the caller enable it explicitly? By doing that, we can still check the uniqueness, if not unique, we can throw exceptions.

Reply:
+1 to an explicit control / visibility

Reply:
Good idea! Added `ParquetInputFormat.COLUMN_ID_RESOLUTION` to control this.